### PR TITLE
CI: enable testing for PT 1.11

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Enable testing for PT 1.11 when it becomes available as wheel for testing

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @akihironitta @carmocca @borda

### CODE REVIEW

**Comment by Borda** [comment]
seems that it was removed from the nighty channel but still was not added to the test channel :rabbit:

**Review by tchaton — APPROVED** [review]
LGTM !

File: .github/workflows/ci_test-conda.yml [hunk 1]
```diff
@@ -18,8 +18,13 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
```
Comment (carmocca) [inline .github/workflows/ci_test-conda.yml:21]: Can we keep this?

As the comment indicates, since "test-full" uses Python 3.9, we use set Python 3.8 for the extra coverage.

File: .github/workflows/ci_test-full.yml [hunk 2]
```diff
@@ -27,7 +27,7 @@ jobs:
         exclude:
           # Skip if torch<1.8 and py3.9 on Linux: https://github.com/pytorch/pytorch/issues/50014
           - {os: ubuntu-20.04, python-version: "3.9", requires: "oldest"}
-        #include:
+        include:
           # nightly: add when there's a release candidate
           #- {os: ubuntu-20.04, python-version: "3.10", requires: "latest", release: "pre"}
```
Comment (carmocca) [inline .github/workflows/ci_test-full.yml:32]: This needs to be included too

```suggestion
          - {os: ubuntu-20.04, python-version: "3.10", requires: "latest", release: "pre"}
```

File: .github/workflows/ci_test-conda.yml [hunk 3]
```diff
@@ -29,10 +34,11 @@ jobs:
       run: |
         conda info
         conda list
+        python .github/prune-packages.py requirements/extra.txt "horovod"
```
Comment (carmocca) [inline .github/workflows/ci_test-conda.yml:37]: Can you add a TODO?

**Review by carmocca — APPROVED** [review]

File: dockers/base-conda/Dockerfile [hunk 4]
```diff
@@ -12,13 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-ARG CUDA_VERSION=11.1
+ARG CUDA_VERSION=11.3.1
 
-FROM nvidia/cuda:${CUDA_VERSION}-devel-ubuntu18.04
+FROM nvidia/cuda:${CUDA_VERSION}-devel-ubuntu20.04
```
Comment (akihironitta) [inline dockers/base-conda/Dockerfile:17]: Shall we keep Ubuntu18.04 for now because of the following error?
```
definition: docker.io/nvidia/cuda:10.2-devel-ubuntu20.04: not found
```
```suggestion
FROM nvidia/cuda:${CUDA_VERSION}-devel-ubuntu18.04
```

File: .github/workflows/ci_test-conda.yml (hunk 3 above)
Comment (Borda) [reply .github/workflows/ci_test-conda.yml:37]:
```suggestion
        # TODO: temporary bypass, need to resolve horovod instalation
        python .github/prune-packages.py requirements/extra.txt "horovod"
```

File: .github/workflows/ci_test-full.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/ci_test-full.yml]: why? troch is not released in stable while it is only testing

File: .github/workflows/ci_test-conda.yml (hunk 1 above)
Comment (Borda) [reply .github/workflows/ci_test-conda.yml:21]: done

File: .github/workflows/ci_test-conda.yml [hunk 5]
```diff
@@ -18,8 +18,13 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
-        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
+        include:
+          - {python-version: "3.8", pytorch-version: "1.7"}
+          - {python-version: "3.8", pytorch-version: "1.8"}
+          - {python-version: "3.8", pytorch-version: "1.9"}
+          - {python-version: "3.8", pytorch-version: "1.10"}
```
Comment (Borda) [inline .github/workflows/ci_test-conda.yml:25]:
```suggestion
        python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
        include:
```

File: .github/workflows/ci_test-full.yml [hunk 6]
```diff
@@ -27,9 +27,9 @@ jobs:
         exclude:
           # Skip if torch<1.8 and py3.9 on Linux: https://github.com/pytorch/pytorch/issues/50014
           - {os: ubuntu-20.04, python-version: "3.9", requires: "oldest"}
-        #include:
-          # nightly: add when there's a release candidate
-          #- {os: ubuntu-20.04, python-version: "3.10", requires: "latest", release: "pre"}
+#        include:
+#          # nightly: add when there's a release candidate
+#          - {os: ubuntu-20.04, python-version: "3.10", requires: "latest", release: "pre"}
```
Comment (Borda) [inline .github/workflows/ci_test-full.yml:32]:
```suggestion
        #include:
          # nightly: add when there's a release candidate
          #- {os: ubuntu-20.04, python-version: "3.10", requires: "latest", release: "pre"}
```

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/ci_test-conda.yml [hunk 7]
```diff
@@ -19,7 +19,11 @@ jobs:
       fail-fast: false
       matrix:
         python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
-        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]  # nightly: add when there's a release candidate
+        pytorch-version: ["1.7", "1.8", "1.9", "1.10"]
+        # nightly: add when there's a release candidate
+        include:
+          # nightly: add when there's a release candidate
```
Comment (carmocca) [inline .github/workflows/ci_test-conda.yml:25]: This is duplicated

```suggestion
```

File: .github/workflows/ci_test-conda.yml [hunk 8]
```diff
@@ -29,10 +33,12 @@ jobs:
       run: |
         conda info
         conda list
+        # TODO: temporary bypass, need to resolve horovod instalation
+        python .github/prune-packages.py requirements/extra.txt "horovod"
```
Comment (carmocca) [inline .github/workflows/ci_test-conda.yml:37]: This should be reverted now that #12234 landed

File: dockers/base-conda/Dockerfile [hunk 9]
```diff
@@ -96,15 +98,16 @@ COPY ./requirements/extra.txt requirements-extra.txt
...
 COPY ./requirements/test.txt requirements-test.txt
 COPY ./requirements/adjust_versions.py requirements_adjust_versions.py
+COPY ./.github/prune-packages.py requirements_prune_packages.py
 
 RUN \
     pip list | grep torch && \
     python -c "import torch; print(torch.__version__)" && \
-    python requirements_adjust_versions.py requirements-extra.txt && \
-    python requirements_adjust_versions.py requirements-examples.txt && \
+    python requirements_prune_packages.py requirements-extra.txt "horovod" "torchtext" && \
```
Comment (carmocca) [inline dockers/base-conda/Dockerfile:106]: This should be reverted now that #12234 landed

File: dockers/base-conda/Dockerfile [hunk 10]
```diff
@@ -129,5 +132,5 @@ RUN \
     conda info && \
     pip list && \
     python -c "import sys; ver = sys.version_info ; assert f'{ver.major}.{ver.minor}' == '$PYTHON_VERSION', ver" && \
-    python -c "import torch; assert torch.__version__.startswith('$PYTORCH_VERSION'), torch.__version__" && \
-    python -c "import horovod.torch"
+    python -c "import torch; assert torch.__version__.startswith('$PYTORCH_VERSION'), torch.__version__"
+    # python -c "import horovod.torch"
```
Comment (carmocca) [inline dockers/base-conda/Dockerfile:136]: This should be reverted now that #12234 landed

File: tests/callbacks/test_finetuning_callback.py [hunk 11]
```diff
@@ -357,24 +358,30 @@ def test_callbacks_restore(tmpdir):
...
+    expected_ = {
         "lr": 0.1,
         "momentum": 0,
         "dampening": 0,
         "weight_decay": 0,
         "nesterov": False,
         "params": ["layer.3.weight", "layer.3.bias"],
     }
+    if _TORCH_GREATER_EQUAL_1_11:
+        expected_.update({"maximize": False})
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:370]:
```suggestion
        expected_["maximize"] = False
```

File: tests/callbacks/test_finetuning_callback.py [hunk 12]
```diff
@@ -357,24 +358,30 @@ def test_callbacks_restore(tmpdir):
     assert len(callback._internal_optimizer_metadata[0]) == 2
 
     # original parameters
-    assert callback._internal_optimizer_metadata[0][0] == {
+    expected_ = {
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:361]: WHy the suffix underscore?

```suggestion
    expected = {
```

File: tests/callbacks/test_finetuning_callback.py [hunk 13]
```diff
@@ -357,24 +358,30 @@ def test_callbacks_restore(tmpdir):
...
         "lr": 0.01,
         "momentum": 0,
         "dampening": 0,
         "weight_decay": 0,
         "nesterov": False,
         "params": ["layer.0.weight", "layer.0.bias"],
     }
+    if _TORCH_GREATER_EQUAL_1_11:
+        expected_.update({"maximize": False})
```
Comment (carmocca) [inline tests/callbacks/test_finetuning_callback.py:383]:
```suggestion
        expected_["maximize"] = False
```

File: tests/callbacks/test_finetuning_callback.py (hunk 11 above)
Comment (Borda) [reply tests/callbacks/test_finetuning_callback.py:370]: any reason?

File: tests/callbacks/test_finetuning_callback.py (hunk 13 above)
Comment (Borda) [reply tests/callbacks/test_finetuning_callback.py:383]: any reason?

File: tests/callbacks/test_finetuning_callback.py (hunk 11 above)
Comment (carmocca) [reply tests/callbacks/test_finetuning_callback.py:370]: Your code creates a new dictionary object unnecessarily when we just need to add a key-value

Comment (Borda) [reply tests/callbacks/test_finetuning_callback.py]: are you sure about it? as the update does not return any new object...

**Comment by Borda** [comment]
@carmocca can you explain why you did 817fc14?
this won't work until they release it or you add the extra path...

File: tests/callbacks/test_finetuning_callback.py (hunk 11 above)
Comment (carmocca) [reply tests/callbacks/test_finetuning_callback.py]: what you are passing to `update` is a newly created dictionary object

**Comment by carmocca** [comment]
Don't we want to also test the prerelease with the "test-full" workflow?

**Comment by Borda** [comment]
> Don't we want to also test the prerelease with the "test-full" workflow?

yes, but then lest fo it and half step to nowhere ;)

File: tests/callbacks/test_finetuning_callback.py (hunk 11 above)
Comment (Borda) [reply tests/callbacks/test_finetuning_callback.py]: ok, now I understand what you mean...

**Comment by Borda** [comment]
ok, seems it is blocked by enabling `horovod>=0.24` in #12248
